consensus/istanbul/backend/announce.go

```diff
@@ -589,130 +649,311 @@ func (sb *Backend) regossipAnnounce(msg *istanbul.Message, payload []byte) error
-func (sb *Backend) handleAnnounceVersionsMsg(peer consensus.Peer, payload []byte) error {
-	logger := sb.logger.New("func", "handleAnnounceVersionsMsg", "peer", peer)
-	var announceVersions []announceVersion
+func (sb *Backend) handleSignedAnnounceVersionsMsg(peer consensus.Peer, payload []byte) error {
+	logger := sb.logger.New("func", "handleSignedAnnounceVersionsMsg")
+	logger.Trace("Handling signed announce version msg")
+	var signedAnnVersions []*signedAnnounceVersion
 
-	if err := rlp.DecodeBytes(payload, &announceVersions); err != nil {
-		logger.Error("Error in decoding announce versions array", "err", err)
+	err := rlp.DecodeBytes(payload, &signedAnnVersions)
+	if err != nil {
+		logger.Warn("Error in decoding received Signed Announce Versions msg", "err", err)
+		return err
 	}
 
-	logger.Trace("Handling an AnnounceVersions message", "announceVersions", fmt.Sprintf("%v", announceVersions))
-
 	// If the announce's valAddress is not within the validator connection set, then ignore it
 	validatorConnSet, err := sb.retrieveValidatorConnSet()
 	if err != nil {
+		logger.Trace("Error in retrieving validator conn set", "err", err)
 		return err
 	}
 
-	announcesToRequest := make([]common.Address, 0)
-	for _, announceVersion := range announceVersions {
-		// Ignore this announceVersion entry if it's val address is not in the current validator connection set.
-		if !validatorConnSet[announceVersion.ValAddress] {
-			logger.Trace("Ignoring announceVersion since it's not in active/elected valset", "valAddress", announceVersion.ValAddress)
+	var validEntries []*vet.SignedAnnounceVersionEntry
+	validAddresses := make(map[common.Address]bool)
+	// Verify all entries are valid and remove duplicates
+	for _, signedAnnVersion := range signedAnnVersions {
+		err := signedAnnVersion.ValidateSignature()
+		if err != nil {
+			logger.Debug("Error validating signed announce version signature", "address", signedAnnVersion.Address, "err", err)
+			continue
+		}
+		if !validatorConnSet[signedAnnVersion.Address] {
+			logger.Debug("Found signed announce version from an address not in the validator conn set", "address", signedAnnVersion.Address)
+			continue
+		}
+		if _, ok := validAddresses[signedAnnVersion.Address]; ok {
+			logger.Debug("Found duplicate signed announce version in message", "address", signedAnnVersion.Address)
 			continue
 		}
+		validAddresses[signedAnnVersion.Address] = true
+		validEntries = append(validEntries, signedAnnVersion.Entry())
+	}
+	return sb.upsertSignedAnnounceVersionEntries(validEntries)
+}
 
-		sb.cachedAnnounceMsgsMu.RLock()
-		if cachedEntry, ok := sb.cachedAnnounceMsgs[announceVersion.ValAddress]; !ok || cachedEntry.MsgVersion < announceVersion.AnnounceMsgVersion {
-			announcesToRequest = append(announcesToRequest, announceVersion.ValAddress)
+func (sb *Backend) upsertSignedAnnounceVersionEntries(entries []*vet.SignedAnnounceVersionEntry) error {
```

Comment:
Nit: I found the contents of this function very surprising given its signature. The actual upsert step takes up 3 lines at the start of the function and the rest is DDoS protection filtering and regossip. Please extract the upsert to the call site and rename this function something more descriptive of its main function, like "regossipAnnounceVersions".